plugins/c8y_firmware_plugin/src/firmware_manager.rs

```diff
@@ -238,25 +268,61 @@ impl FirmwareManager {
         smartrest_request: SmartRestFirmwareRequest,
         operation_id: String,
     ) -> Result<(), FirmwareManagementError> {
-        let child_id = smartrest_request.device.as_str();
         let firmware_url = smartrest_request.url.as_str();
         let file_cache_key = digest(firmware_url);
-        let file_transfer_url = format!(
-            "http://{}/tedge/file-transfer/{child_id}/firmware_update/{file_cache_key}",
-            &self.local_http_host
-        );
 
         // <tedge-persistent-root>/cache/<file_cache_key>
         let cache_dest_path = self.get_cache_file_path(&file_cache_key)?;
 
-        self.download_new_file(&cache_dest_path, firmware_url)
+        if cache_dest_path.is_file() {
+            info!(
+                "Hit the file cache={}. File download is skipped.",
+                cache_dest_path.display()
+            );
+            self.handle_firmware_update_request_with_downloaded_file(
+                smartrest_request,
+                operation_id,
+                cache_dest_path,
+            )
             .await?;
+        } else {
+            let download_req = DownloadRequest {
+                url: firmware_url.into(),
+                file_name: file_cache_key,
+            };
+
+            info!("Downloading the firmware from {}", firmware_url);
+            // Send a request to the DownloadManager to download the file asynchornously
+            self.download_req_sndr.send(download_req).await?;
+            self.req_in_progress = Some((smartrest_request.clone(), operation_id));
+        }
+
+        Ok(())
+    }
+
+    async fn handle_firmware_update_request_with_downloaded_file(
+        &mut self,
+        smartrest_request: SmartRestFirmwareRequest,
+        operation_id: String,
+        downloaded_firmware: PathBuf,
+    ) -> Result<(), FirmwareManagementError> {
+        let child_id = smartrest_request.device.as_str();
+        let firmware_url = smartrest_request.url.as_str();
+        let file_cache_key = digest(firmware_url);
+
+        // <tedge-persistent-root>/cache/<file_cache_key>
+        let cache_dest_path = self.get_cache_file_path(&file_cache_key)?;
+        move_file(&downloaded_firmware, &cache_dest_path)?;
```

Comment:
When hitting in the cache, it is weird to call this `move_file()` here. This line is meant to move a file from `tmp_dir` to `/var/tedge/cache`.

Reply:
Currently it's harmless as the source and destination would be the same and hence nothing would happen. But yeah, this could be avoided, but I'd need some more refactoring to avoid this. Will try to do it in a follow-up PR if it's too complex.

Reply:
Resolved by adding an explicit path check before moving.